### ISSUE TITLE
Secrets rotation and Secrets semantics

### ISSUE THREAD

**Issue by erictune** [GoogleCloudPlatform/kubernetes#4673]
It would be nice to be able to rotate secrets while the pod is still running.  This issue is to discuss how that might work.

Some things to think about:
  - Can we attach a new volume to a running container?  I don't think so.
  - Is the content of a secret mutable after creation?
  - What happens if you delete a secret that is in use by running pods?  I guess they keep running, but then new instances can't be created from the same template?
  - What happens if you modify the data in a pod (if that is allowed) or if you delete a secret and then create a new one with the same name?   Does kubelet notice that and update the volume of a running pod?  
  - Given that the tmpfs is readonly bind mounted, if the kubelet could modify the data, how consistently would the pod see the changes?
  - If the kubelet does modify volume contents, should it do so in a way that is atomic (that is, make a second file tree, and then flip a hard link over once the change is complete?  Does that require a minimum number of path components to separate the bind mount root from the flippable hard link?
  - If the above, then we should document that clients need to not keep secrets files open indefinitely.
  - If containers keep files open after files on the tmpfs are deleted, the memory charges _should_ migrate to the container (@vmarmol or @dchen1107 or @thockin might know better)
  - Is the behavior different for different secrets types?

**Comment by erictune** [comment]
Relevant to #4672

**Comment by erictune** [comment]
@pmorie

**Comment by erictune** [comment]
@bgrant0607

**Comment by thockin** [comment]
I'm really hoping to avoid having the kubelet be dynamically updating magic
volumes (run your own side-car), but I am not sure we can escape it with
secrets since there is info there that is meaningful to the kubelet itself..

**Comment by bgrant0607** [comment]
I think we should support dynamic updates. That makes the feature vastly more useful and reduces the complexity for clients tremendously. Immutable secret objects would induce cascading complexity: automatic name generation, automatic garbage collection of secrets not in use, etc. Plus, I doubt we could prevent users from updating the volume source to point to a new secret, which seems to have about the same complexity. In general, making things not updatable greatly complicates our user-facing configuration solution. We could potentially restart/recreate containers upon these kinds of updates for now, if that would simplify the implementation.

Additionally, we're going to want the same behavior for a dynamic config solution for our own components.

**Comment by bgrant0607** [comment]
Also, I vote for atomic update. A lifecycle event can notify the containers that they should reopen the file(s).

**Comment by thockin** [comment]
Dynamic updates can be achieved by running a secret-syncer sidecar.  I'm
just trying to convince myself that this is not a viable answer.

As for atomic updates, easier said than done.  The only atomic op is
rename() and that is singular AND only works for directories iff the
destination does not already exist.

We could do something like put secrets into a subdir of the volume and then
flip a symlink with rename.

e.g.

mkdir /path/to/volume/.update_in_progress
write secrets to /path/to/volume/.update_in_progress
mv /path/to/volume/.update_in_progress /path/to/volume/new_timestamp
ln -s new_timestamp /path/to/volume/.updated_link
rename /path/to/volume/.updated_link /path/to/volume/current

It's ugly and we still have to deal with cleanup, but it's the only way I
know.

**Comment by pmorie** [comment]
@erictune @thockin @bgrant0607 I am torn on this one personally.  One thing I want to point out while I gather my thoughts is that I think we have a similar problem in this use-case to other ones that have been mentioned:

- Expose labels to containers (#560)
- Dynamic configuration (#1553)
- Exposing secrets as env vars (#4710)

**Comment by erictune** [comment]
I now think we probably don't need to prioritize automated rotation at this time.  
 
Most pods are going to have a way that they can be updated to have a new image or new flags.  Updating the secret can use that same workflow, assuming secret changes are infrequent.
 
There are two cases where secrets might change frequently, but I don't think these are that important right now:

First, short lived tokens.  We've talked about having a way to hold something like an OAuth Refresh Token in the apiserver, and then have it generate access tokens periodically, and "rotate" them for clients.  While this has the nice property of protecting the long lived secret, clients are probably not going to expect to have the access token changing underneath them, and it has the drawback of giving the client the access token even when it doesn't need it (if it has time-varying behavior).

Second would be a server that stores multiple changing secrets, like username/password combinations for a website.  This case doesn't match the current secrets implementation: secret modifications made by the pod won't be remembered.

**Comment by erictune** [comment]
TL;DR
I don't think we should work on secret rotation right now.
And we should make secrets immutable.  I'll file a narrower issue for that.

**Comment by thockin** [comment]
I'm fine closing this for now - it's a much larger design issue and not priortiy
